internal/http/parser/http_response.go

```diff
@@ -0,0 +1,225 @@
+package http_parser
+
+import (
+	"bytes"
+	"fmt"
+	"io"
+	"strconv"
+	"strings"
+)
+
+type Header struct {
+	Key   string
+	Value string
+}
+
+type StatusLine struct {
+	Version    string
+	StatusCode int
+	Reason     string
+}
+
+type HTTPResponse struct {
+	StatusLine StatusLine
+
+	Headers []Header
+
+	Body []byte
+}
+
+func Parse(data []byte) (httpResponse HTTPResponse, readBytesCount int, err error) {
+	reader := bytes.NewReader(data)
+
+	response, err := doParseResponse(reader)
+	if err != nil {
+		return HTTPResponse{}, 0, err
+	}
+
+	return response, len(data) - reader.Len(), nil
+}
+
+func doParseResponse(reader *bytes.Reader) (HTTPResponse, error) {
```

Comment:
@ryan-gang this is a pretty long function, based on the spec I think the cleanest way to structure this would be to have three high-level functions that you call: 

`parseStatusLine`, `parseHeaders` and `parseBody`

There's a lot of imperative stuff going on here like `allHeadersFound`, `sectionsFound` etc. All of this can be avoided if you instead pass the `reader` to functions and just let them either parse or return an error.